### PR TITLE
Inter mesh projection

### PR DESCRIPTION
**Pull request by vikramvgarg**
This PR adds functionality to project vectors from one mesh to another arbitrary mesh via the new InterMeshProjection class. The approximation basis used in either system can be arbitrary including C1 elements.

InterMeshProjection objects will take in 'to' and 'from' system references, and project vectors between them. To handle the C1 case a GradientMeshFunction class derived from FunctionBase<Gradient> has been provided. This class includes a shim function and lets System::project_vector take a gradient evaluation functor along with a MeshFunction for point values.

Subsequent branch adding MeshHistory (the mesh analog of SolutionHistory) will use InterMeshProjection objects to project adjoint solution vectors from one adjoint timestep to another, where each timestep can have a different mesh.

Manually tested on several cases:
1) Same geometry
2) h-Refined and h-coarsened mesh test
3) Adjoint vector projection test.
4) Multi variable test.
5) 2-dimensional test.
6) C1 elements in the to_system.
7) Parallel test.

### CODE REVIEW

File: Makefile.in [hunk 1]
```diff
@@ -524,6 +524,7 @@ am__libmesh_dbg_la_SOURCES_DIST = src/base/default_coupling.C \
 	src/systems/explicit_system.C src/systems/fem_context.C \
```
Comment (roystgnr) [inline Makefile.in:524]: Makefile.in (and configure, and anything else generated by bootstrap) changes need to be in a separate commit from all other changes; and that commit needs to be after the Makefile.am changes that led to them, not before.

File: include/systems/inter_mesh_projection.h [hunk 2]
```diff
@@ -77,7 +77,9 @@ namespace libMesh
         {mesh_function = _mesh_function;}
 
         // Destructor
-        virtual ~GradientFunction () {}
+        virtual ~GradientFunction () { }
+
+        virtual void init () { mesh_function->init(); }
```
Comment (roystgnr) [inline include/systems/inter_mesh_projection.h:86]: This whole commit looks like it ought to be a fixup commit to 2f9f8cf9c

File: src/systems/inter_mesh_projection.C [hunk 3]
```diff
@@ -18,14 +18,25 @@
...
+#include "libmesh/enum_norm_type.h"
+
+
 // C++ includes
 
 namespace libMesh
 {
+    GradientFunction::GradientFunction(MeshFunction * _mesh_function)
+    {
+     //mesh_function = std::unique_ptr<MeshFunction>(dynamic_cast<MeshFunction *>((_mesh_function->clone()).get()));
```
Comment (roystgnr) [inline src/systems/inter_mesh_projection.C]: You were on a better track with this one.  We know we will always have a MeshFunction in mesh_function, so let's do that, then we can do the casting just once, when the unique_ptr is assigned, not every single time it's used.  Let's also use libmesh_cast_ptr, so it's not doing a dynamic_cast in opt mode or on systems where RTTI is disabled.

File: src/systems/inter_mesh_projection.C [hunk 4]
```diff
@@ -38,27 +49,144 @@ namespace libMesh
...
         unsigned int n_vars = from_system.n_vars();
         libmesh_assert_equal_to (to_system.n_vars(), n_vars);
 
+        // We are going to use the multi-variable MeshFunction, so we can pass
+        // a single vector of variables rather than have a MeshFunction for each variable
+	    std::vector<unsigned int> variables_vector;
+
+        for (unsigned int j = 0; j != n_vars; ++j)
+        {
+          variables_vector.push_back(j);
```
Comment (roystgnr) [inline src/systems/inter_mesh_projection.C:63]: Let's add an assertion here that the names match for each variable?  It's unlikely, but I could imagine someone adding variables to the to-system in a different order than they were added to the from-system and then getting burned.

File: src/systems/inter_mesh_projection.C [hunk 5]
```diff
@@ -38,27 +49,144 @@ namespace libMesh
...
         from_system.update();
 
-        // Loop over all the vectors in the system
-        for (unsigned int i = 0; i != n_vecs; ++i)
-        {
+        // Any system holds the solution along with the other vectors system.vectors
+        // We will first project the solution and then move to the system.vectors
+
+        // Construct local version of the current system vector
+	    // This has to be a serial vector
```
Comment (roystgnr) [inline src/systems/inter_mesh_projection.C]: Technically it just has to be a ghosted vector whose algebraically ghosted values cover a domain which is a superset of the to-system's domain ... that's hard to do and we can skip it now, but we might want to put a "FIXME" comment or something to that effect so we don't simply forget about the problem until the poor scalability bites someone.

File: src/systems/inter_mesh_projection.C [hunk 6]
```diff
@@ -38,27 +49,144 @@ namespace libMesh
...
+        // Print the vector
+        for(unsigned int i = 0; i < from_system.solution->size(); i++)
+            std::cout<<(*from_system.solution)(i)<<std::endl;
+
+        std::cout << "T_from(H1) = " << T_from_H1 << std::endl << "T_from(L2) = " << T_from_L2 << std::endl;
+
+        // End testing
+
+        // FIXME: To properly apply adjoint Dirichlet constraints, we need to pass a fourth argument when we are projecting adjoint vectors.
+        to_system.project_vector(*to_system.solution, dynamic_cast<FunctionBase<Number> *>(mesh_func_solution), dynamic_cast<FunctionBase<Gradient> *>(gptr_solution));
```
Comment (roystgnr) [inline src/systems/inter_mesh_projection.C:91]: Why do you have dynamic_cast here?  C++ doesn't need any sort of casting to interpret a subclass pointer as a base class pointer.

File: include/systems/inter_mesh_projection.h [hunk 7]
```diff
@@ -42,7 +42,7 @@ namespace libMesh
     /**
      * This class implements inter mesh projection, i.e. projection of
      * vectors defined on a given mesh (from_mesh associated with from_system)
-     * to another mesh (to_mesh), stored in the map projected_vectors_map.
+     * to another mesh (to_mesh of to_system).
```
Comment (roystgnr) [inline include/systems/inter_mesh_projection.h:45]: This is another "probably ought to just be a fixup" commit.  We like having lots of intermediate working commits in the git history, because that helps pinpoint where breakage occurred when a new test fails, but broken commits don't help bisect anything.

I also wouldn't say "further testing completed" at this point; manually run test results aren't the same as unit tests.  A completed test is one that gets rerun with every "make check"; it's too easy to regress otherwise.

File: src/systems/inter_mesh_projection.C (hunk 3 above)
Comment (vikramvgarg) [reply src/systems/inter_mesh_projection.C]: Yes, that does make sense.

Comment (vikramvgarg) [reply src/systems/inter_mesh_projection.C]: Fixed in an upcoming commit.

File: src/systems/inter_mesh_projection.C (hunk 4 above)
Comment (vikramvgarg) [reply src/systems/inter_mesh_projection.C:63]: Fixed in an upcoming commit.

File: src/systems/inter_mesh_projection.C (hunk 5 above)
Comment (vikramvgarg) [reply src/systems/inter_mesh_projection.C]: Added in upcoming commit.

File: src/systems/inter_mesh_projection.C (hunk 6 above)
Comment (vikramvgarg) [reply src/systems/inter_mesh_projection.C:91]: Fixed in upcoming commit.

**Comment by roystgnr** [comment]
This looks good now; Vikram's going to do the interactive rebase and if there are no problems there we'll merge after the rebased branch passes CI.

**Comment by roystgnr** [comment]
6b917f7 probably should have been a fixup on top of 8878e61, but I think we're at the point where speculative additional difficulty bisecting bugs in the future is less than the definite additional difficulty futzing with things more now.  I'll merge if/when that last kicked test is happy.

**Review by roystgnr — APPROVED** [review]
